### ISSUE TITLE
DoH almost requires let's encrypt integration

### ISSUE THREAD

**Issue by miekg** [coredns/coredns#1827]
With experimental DOH support merged, we will be (almost) forced to think about Let's Encrypt integration; I have no idea how much work that is; hopefully we can ~~rip off~~ copy Caddy here, or better yet outsource this (pkg wise) entirely. 

Having the certs sit some where also requires the concept of local storage (isn't a first class citizen in CoreDNS), so that needs some thought.

Another thing I didn't like in Caddy's orginal impl. was that a cluster of Caddy's couldn't easily share the certs; they made that better by having a some kind of idealized shared storage (IIRC). I would like to use something that requires no storage. Maybe some kind of master selection and cert distribution using gRPC; but having this requires a discovery mechanism as well; all these things are extremely hard to execute on; let alone coming up with a robust implementation. We should probably piggy back on some storage backend (sql, etcd, or even disk).

**Comment by m4r10k** [comment]
Currently the golang supports more or less the ACME process for the certificate management, but of course does not provide a "full" implementation (https://godoc.org/golang.org/x/crypto/acme/autocert) This means, that if CoreDNS would like to provide LetsEncrypt certificates in the core, it has to provide the different mechanisms and there abstractions on getting and renewing certificates like DNS( AWS, ...), http-Hooks, ... too.

In my opinion, this is hard to achieve. Beside this, putting a DNS server on the internet to retrieve certificates for DOH is not that secure.

There are already good implementations for the ACME things which are working really well. We use CERTBOT for example, to retrieve and renew the certificates with different protocols (dns, http). So it should be easy to achieve this task and in the end you get a certificates somewhere.

So it would be better to just pick up the certificates at some point - etcd might be one way, but in kubernetes you would use a configmap maybe and therefore a filepath or like we do it, we store them in a Git and use a sidecar for it because in the end, you have the problem, that you need to reload your http implementation so that the new certificate can be picked up.

Therefore, for the container world, I included a simple go cron library in our load balancer project, which just reloads the service once a day (no restart, reload) to pick up the new certs, if there are some.

It's simple and reliable, because the CERTBOT begins to start the renewal of the certificate 30 days before expiration. So there is more than enough time to pick up the new one.

To sum up: It would be nice to have a LetsEncrypt plugin which can be configured to load the certs from certain sources like etcd, file, whatever... maybe.

**Comment by kfox1111** [comment]
One option is to just implement http rather then https. For sites deploying in k8s, cert manager can be used along with let's encrypt and an ingress controller. All of that plumbing can deal with the https termination.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] DoH almost re..." ]
>One option is to just implement http rather then https. For sites deploying in k8s, cert manager can be used along with let's encrypt and an ingress controller. All of that plumbing can deal with the https termination.

Yeah, out sourcing this seems to be the best option, various k8s things need 
this more than coredns and it will be interesting to see what comes out of this.

**Comment by jameshartig** [comment]
Yeah, seems much easier to allow a reverse proxy to sit in front of coredns for now.

**Comment by m4r10k** [comment]
Yes, it is!

**Comment by miekg** [comment]
Closing as rev. proxy seems to be a way forward.
